File: code/src/lib/Bvr/version_vault/core_old.cpp
#include <cstdlib>
#include <cstdio>
#include <string>
#include <vector>
#include <iostream>

using namespace std;

bool runCommand(const string& command) {
    return system(command.c_str()) == 0;
}

string getCommandOutput(const string& command) {
    FILE* pipe = popen(command.c_str(), "r");

    if (!pipe)
        return "";

    char buffer[256];
    string result;

    while (fgets(buffer, sizeof(buffer), pipe))
        result += buffer;

    pclose(pipe);

    if (!result.empty() && result.back() == '\n')
        result.pop_back();

    return result;
}

bool initRepo(const string& repoPath) {

    string command =
        "git init \"" + repoPath + "\"";

    return runCommand(command);
}

bool trackFile(
    const string& repoPath,
    const string& fileId,
    const string& filePath
) {

    string command =
        "cd \"" + repoPath + "\" && "
        "git checkout --orphan file-" + fileId;

    if (!runCommand(command))
        return false;

    command =
        "cp \"" + filePath + "\" \"" +
        repoPath + "/tracked_file\"";

    if (!runCommand(command))
        return false;

    command =
        "cd \"" + repoPath + "\" && "
        "git add tracked_file";

    if (!runCommand(command))
        return false;

    command =
        "cd \"" + repoPath + "\" && "
        "git commit -m \"Version 1\"";

    if (!runCommand(command))
        return false;

    command =
        "rm \"" + repoPath + "/tracked_file\"";

    if (!runCommand(command))
        return false;

    return true;
}

bool saveVersion(
    const string& repoPath,
    const string& fileId,
    const string& filePath
) {

    string command =
        "cd \"" + repoPath + "\" && "
        "git checkout file-" + fileId;

    if (!runCommand(command))
        return false;

    command =
        "cp \"" + filePath + "\" \"" +
        repoPath + "/tracked_file\"";

    if (!runCommand(command))
        return false;

    command =
        "cd \"" + repoPath + "\" && "
        "git add tracked_file";

    if (!runCommand(command))
        return false;

    command =
        "cd \"" + repoPath + "\" && "
        "git commit -m \"New Version\"";

    if (!runCommand(command))
        return false;

    command =
        "rm \"" + repoPath + "/tracked_file\"";

    if (!runCommand(command))
        return false;

    return true;
}

vector<string> getHistory(
    const string& repoPath,
    const string& fileId
) {

    string command =
        "cd \"" + repoPath + "\" && "
        "git log --format=%H refs/heads/file-" +
        fileId;

    FILE* pipe = popen(command.c_str(), "r");

    vector<string> history;

    if (!pipe)
        return history;

    char buffer[256];

    while (fgets(buffer, sizeof(buffer), pipe)) {

        string commit = buffer;

        if (!commit.empty() &&
            commit.back() == '\n') {

            commit.pop_back();
        }

        history.push_back(commit);
    }

    pclose(pipe);

    return history;
}

bool restoreVersion(
    const string& repoPath,
    const string& fileId,
    const string& commitHash,
    const string& filePath
) {

    string command =
        "cd \"" + repoPath + "\" && "
        "git checkout " + commitHash +
        " -- tracked_file";

    if (!runCommand(command))
        return false;

    command =
        "cp \"" + repoPath + "/tracked_file\" \"" +
        filePath + "\"";

    if (!runCommand(command))
        return false;

    command =
        "cd \"" + repoPath + "\" && "
        "git restore --staged tracked_file";

    runCommand(command);

    command =
        "rm \"" + repoPath + "/tracked_file\"";

    runCommand(command);

    return true;
}

File: code/src/lib/Bvr/version_vault/core_old.py

import os
import shutil
import subprocess


def runCommand(command):
    return subprocess.run(command, shell=True).returncode == 0


def getCommandOutput(command):
    result = subprocess.run(
        command,
        shell=True,
        capture_output=True,
        text=True
    )

    if result.returncode != 0:
        return ""

    return result.stdout.rstrip("\n")


def initRepo(repoPath):
    command = f'git init "{repoPath}"'
    return runCommand(command)


def trackFile(repoPath, fileId, filePath):
    command = (
        f'cd "{repoPath}" && '
        f'git checkout --orphan file-{fileId}'
    )

    if not runCommand(command):
        return False

    command = (
        f'cp "{filePath}" '
        f'"{repoPath}/tracked_file"'
    )

    if not runCommand(command):
        return False

    command = (
        f'cd "{repoPath}" && '
        f'git add tracked_file'
    )

    if not runCommand(command):
        return False

    command = (
        f'cd "{repoPath}" && '
        f'git commit -m "Version 1"'
    )

    if not runCommand(command):
        return False

    command = f'rm "{repoPath}/tracked_file"'

    if not runCommand(command):
        return False

    return True


def saveVersion(repoPath, fileId, filePath):
    command = (
        f'cd "{repoPath}" && '
        f'git checkout file-{fileId}'
    )

    if not runCommand(command):
        return False

    command = (
        f'cp "{filePath}" '
        f'"{repoPath}/tracked_file"'
    )

    if not runCommand(command):
        return False

    command = (
        f'cd "{repoPath}" && '
        f'git add tracked_file'
    )

    if not runCommand(command):
        return False

    command = (
        f'cd "{repoPath}" && '
        f'git commit -m "New Version"'
    )

    if not runCommand(command):
        return False

    command = f'rm "{repoPath}/tracked_file"'

    if not runCommand(command):
        return False

    return True


def getHistory(repoPath, fileId):
    command = (
        f'cd "{repoPath}" && '
        f'git log --format=%H refs/heads/file-{fileId}'
    )

    result = subprocess.run(
        command,
        shell=True,
        capture_output=True,
        text=True
    )

    if result.returncode != 0:
        return []

    return result.stdout.splitlines()


def restoreVersion(repoPath, fileId, commitHash, filePath):
    command = (
        f'cd "{repoPath}" && '
        f'git checkout {commitHash} -- tracked_file'
    )

    if not runCommand(command):
        return False

    command = (
        f'cp "{repoPath}/tracked_file" '
        f'"{filePath}"'
    )

    if not runCommand(command):
        return False

    command = (
        f'cd "{repoPath}" && '
        f'git restore --staged tracked_file'
    )

    runCommand(command)

    command = f'rm "{repoPath}/tracked_file"'
    runCommand(command)

    return True

